### PR TITLE
add verify_assert() and remove side-effects from existing assert() usage

### PR DESCRIPTION
**Pull request by kv2019i**
A patch to bring SOF closer to standard C99 usage of assert.h.

This cleans up the code, and gives the option to squeeze out a little bit more performance by building with all assert()s disabled (not the default but at least now possible).

A draft pull-req as needs still a bit more testing (and double-check I got all the assert misusage).

### CODE REVIEW

File: src/include/sof/debug/panic.h [hunk 1]
```diff
@@ -32,14 +32,26 @@ void panic_dump(uint32_t p, struct sof_ipc_panic_info *panic_info,
...
-	do {			\
-		if (!(x))	\
-			k_oops();\
+	do { \
+		__ASSERT_NO_MSG(x); \
 	} while (0)
-#else
+#else /* !__ASSERT_ON */
+#define assert(x) (void)(x)
+#endif
```
Comment (lyakh) [inline src/include/sof/debug/panic.h:52]: You're saying that currently SOF's usage of `assert()` has side-effects, so if we compile it out, unexpected things might happen, right? So, to fix this we first have to eliminate those potential side effects, and just disabling `assert()` could break things. But here you make `assert()` depend on `__ASSERT_ON` which isn't defined in our current builds, so this patch already disables `assert()`? Am I missing something?

File: src/audio/mux/mux.c [hunk 2]
```diff
@@ -300,8 +300,9 @@ static int mux_ctrl_get_cmd(struct comp_dev *dev,
 			sizeof(struct mux_stream_data);
 
 		/* copy back to user space */
-		assert(!memcpy_s(cdata->data->data, ((struct sof_abi_hdr *)
-				 (cdata->data))->size, cfg, reply_size));
+		verify_assert(!memcpy_s(cdata->data->data,
+					((struct sof_abi_hdr *)(cdata->data))->size,
+					cfg, reply_size));
```
Comment (lyakh) [inline src/audio/mux/mux.c:305]: this breaks bisect...

**Review by lyakh — CHANGES_REQUESTED** [review]
I think the order of changes is wrong: the first patch (potentially) breaks code and following patches attempt to fix it again

File: src/include/sof/debug/panic.h (hunk 1 above)
Comment (kv2019i) [reply src/include/sof/debug/panic.h:52]: Ack @lyakh , I'll add more comments. Intention is definitely to keep bisect working. In short with this patch, the assert expression is always evaluated:
 case A) If you have Zephyr's __ASSERTs enabled (CONFIG_ASSERT=y and friends), we call "__ASSERT_NO_MSG(x)" and code is generated to evaluate (x).
  case B) If you have __ASSERT==0 (e.g. no code is produced for "__ASSERT_NO_MSG(x)"), we emit the expression to compiler output directly with "(void)(x)".

So e.g. "assert(memcpy())" always keeps the memcpy() in compiler output.

**Comment by plbossart** [comment]
asserts should not be used, period. I've had so many issues in my past with them, it's one of those 'hell is paved with good intentions' complete nightmare.

**Comment by marc-hb** [comment]
So `assert(!memcpy_s())` is a simple bug, thanks for trying to fix it. However I don't think this requires the complexity of new macros with new conditionals. What's wrong with this?

```C

errno_t memcpy_err __unused = memcpy_s(....);
assert(memcpy_err);

```
Keep it simple! Submitted in PR #5758 and seems to work.

-----

`assert()` semantics are very clear: because asserts are optional, its argument should never have any side-effect[*], period. `assert()` arguments should be treated exactly the same as logging arguments. Simple.


Granted: it's not always clear _when_ asserts are enabled and when they are not. That does not matter and again it's exactly the same for log statements. In general, no one understands anything to Kconfig or any other build configuration knob anyway and no one can ever predict what exactly gets compiled, nothing specific to `assert` here (but this new `#if __ASSERT_ON` does not help)
For instance who could have predicted that our entire validation suddenly started to test with `CONFIG_HYPERVISOR_GUEST`, `CONFIG_KVM_GUEST` and what else suddenly turned ON? https://github.com/thesofproject/kconfig/pull/70. No one; configuration control is a long lost cause.

[*] If some developer struggles with the concept of a "side-effect" then it means that (among other issues) they can't even use logging statements which means they can't be allowed to participate in any professional software project yet.

**Comment by kv2019i** [comment]
Let's go with #5758 instead.

**Pull request closed without merging**